### PR TITLE
Metadata Viewer: API to return dict, not list

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to address the request, "File metadata extracted from the API is hard to parse. Maybe a dictionary is better?"

I avoided changing `metadata` itself because that would mean messing with the front-end stuff that already works. Also might be subtly confusing because the same attribute now returns something different. Maybe best to have a clean break using a different attribute name.

### Change log entry

- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`? If you want to avoid merge conflicts,
  list the proposed change log here for review and add to `CHANGES.rst` before merge. If no, maintainer
  should add a `no-changelog-entry-needed` label.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a milestone set? Set this to bugfix milestone if this is a bug fix and needs to be released ASAP; otherwise, set this to the next major release milestone. Bugfix milestone also needs an accompanying backport label.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)? [🐱](https://jira.stsci.edu/browse/JDAT-4930)

### CODE REVIEW

File: jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py [hunk 1]
```diff
@@ -101,7 +103,8 @@ def find_public_metadata(self, meta, primary_only=False):
         d = flatten_nested_dict(meta)
         # Some FITS keywords cause "# ipykernel cannot clean for JSON" messages.
         # Also, we want to hide internal metadata that starts with underscore.
-        badkeys = ['COMMENT', 'HISTORY', ''] + [k for k in d if k.startswith('_')]
+        badkeys = (['COMMENT', 'HISTORY', ''] + [k for k in d if k.startswith('_')]
+                   + [k for k in d if k.startswith('original_')])
```
Comment (pllim) [inline jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:107]: Without this, I get things in Cubeviz that is not from original header like:

* original_spectral_axis_unit (a Quantity)
* original_wcs (WCS object)

Perhaps these should have been hidden with leading underscore but they snuck through so here we are.

Comment (pllim) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:107]: Weird thing is I don't know where these are being injected. A simple grep in the code base gave me no match.

File: jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py [hunk 2]
```diff
@@ -47,14 +48,15 @@ def __init__(self, *args, **kwargs):
 
     @property
     def user_api(self):
-        return PluginUserApi(self, expose=('dataset', 'show_primary'), readonly=('metadata',))
+        return PluginUserApi(self, expose=('dataset', 'show_primary'), readonly=('meta',))
```
Comment (kecnry) [inline jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:51]: `metadata` was public API, so maybe we should keep and deprecate?  Is the change in name only because the change in format?

Comment (pllim) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:51]: > deprecate

Perhaps. Depends on whether people insist of changing `metadata` type itself or not during review. I will hold off until we have agreement. Thanks.

Comment (kecnry) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:51]: > metadata as a list is used in Vue for actual displaying as well. Changing metadata itself from list to dict would require more refactoring on the front end.

The only potential difficult part might be sorting, but OrderedDict might do the trick if traitlets accept them.

But we can also always have the internal attribute renamed to something just for internal use (`metadata_list` or something).

Comment (pllim) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:51]: There is a `Dict` traitlet but I don't know exactly how it is going to behave. Because we also grab the header comments, we still need the list for front end anyway, so `metadata_list` might be the way to go. Still need more refactoring though. I can look into that if you prefer that over deprecating `metadata` from public API. Lemme know.

**Comment by pllim** [comment]
During 2024-11-15 tag-up, it was agreed to deprecate `metadata` as public API.

> carry also the description

@camipacifici , this was also discussed at tag-up but unclear if this will be confusing to users or not. Description is only applicable when it comes from FITS header (but even then, not always there). In `astropy.io.fits.Header`, such things are represented by `key: (value, descrip)` combo but a user that is not familiar with that convention might not understand why the key gives back a tuple.

@kecnry suggested a nested dictionary. But will a nested dictionary be confusing too? Also a nested dictionary cannot be roundtripped back into a FITS Header easily.

**Comment by kecnry** [comment]
nested dictionary could look something like:

```
{'ACT_ID': {'description': 'Activity identifier', 'value': 1},
 'ASPERNAME': {'description': 'PRD science aperture used': 'value': 'NRS_FULL_MSA'}
...
}
```

**Comment by rosteen** [comment]
> Also a nested dictionary cannot be roundtripped back into a FITS Header easily.

Is there somewhere in the code that we do this in Jdaviz that would be made more complicated, or are you talking about a user who might want to put the metadata returned by this back into a Header?

**Comment by pllim** [comment]
> are you talking about a user who might want to put the metadata returned by this back into a Header?

Yes, there is always a chance a user did a cube smooth or whatever and then wants to write their own FITS file back out and then want to stuff in header from original cube.

**Comment by pllim** [comment]
Okay, looks like I give `fits.Header` too much credit. This works if `d` is a simple key-value pair.

```python
fits.Header({"a": 1})
```

This does not work.

```python
fits.Header({"a": (1, "aaa")})
```

But if you do it keyword by keyword, this works.

```python
hdr = fits.Header()
hdr["a"] = (1, "aaa")
```

So I guess simple full roundtrip is quite impossible. Just a matter of how much code you want user to write, like this.

```python
for key, val in meta.items():
    hdr[key] = val
```

Or like this.

```python
for key, val in meta.items():
    hdr[key] = (val["value"], val.get("description", ""))
```

Any preference?

**Comment by rosteen** [comment]
I'm in favor of the nested dictionary approach, I think that's the clearest thing to return. We could always provide a `meta_as_header` method in the API to do the fiddly bits of packing it back up in a FITS header (as a follow up, I don't think that would need to go in this PR).

File: jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py (hunk 1 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:107]: Oh yeah, I think @kecnry or @rosteen was right. These are from specutils

https://github.com/astropy/specutils/blob/170d70f4d3d0e9db624b1346ae97b5ad5f7b2fbe/specutils/spectra/spectrum_mixin.py#L243-L248

File: jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py [hunk 3]
```diff
@@ -47,7 +47,11 @@ def __init__(self, *args, **kwargs):
 
     @property
     def user_api(self):
-        return PluginUserApi(self, expose=('dataset', 'show_primary'), readonly=('metadata',))
+        plg_api = PluginUserApi(self, expose=('dataset', 'show_primary'),
+                                readonly=('metadata', 'meta',))
+        plg_api._deprecation_msg = ("in the future, \"metadata\" will not be publicly accessible, "
```
Comment (pllim) [inline jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:52]: I don't think this is correct but I cannot find a good example of deprecating just an attribute that is also traitlet. Just so we're clear, we are never going to delete this, just retiring it from *public* API. Maybe @kecnry can tell me how when he returns next week.

Comment (kecnry) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py:52]: right, this is intended when the entire plugin (or plugin name) is deprecated.  I don't know if we've had a case with an attribute before, so we can either write logic into `PluginUserAPI` (maybe a separate `deprecated` list that acts similar to `readonly` but raising the warnings), or just make `meta` a property here and use the deprecated decorator.

File: jdaviz/core/user_api.py [hunk 4]
```diff
@@ -34,6 +35,9 @@ def __getattr__(self, attr):
         if attr in _internal_attrs or attr not in self._expose:
             return super().__getattribute__(attr)
 
+        if attr in self._deprecated:
+            logging.warning("DeprecationWarning: %s is deprecated" % attr)
```
Comment (pllim) [inline jdaviz/core/user_api.py:39]: Screenshot 2024-11-18 114104

File: jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py (hunk 3 above)
Comment (pllim) [reply jdaviz/configs/default/plugins/metadata_viewer/metadata_viewer.py]: OK I reverted this and added something for attr in the API parent class.

File: jdaviz/core/user_api.py (hunk 4 above)
Comment (pllim) [reply jdaviz/core/user_api.py:39]: I thought about making it fancier so it can say "use meta instead" but decided against it because hopefully we won't do this often and I don't want to overcomplicate the machinery.

**Comment by pllim** [comment]
Thanks for the reviews!
